provider/block_devices_plug.py

```diff
@@ -0,0 +1,436 @@
+        for i in xrange(max_threads):
+            mon = self._vm.monitors[i % (len(self._vm.monitors) - 1)]
+            args = (self._vm, action, imgs[i::max_threads], mon, self.exit_event)
+            if bus:
+                args = (self._vm, action, imgs[i::max_threads], mon, self.exit_event, bus)
+            self._threads.append(_PlugThread(*args))
+
+    def _start_threads(self):
+        for thread in self._threads:
+            thread.start()
+
+    def _join_threads(self, timeout):
+        for thread in self._threads:
+            thread.join(timeout)
+
+    def run_threads(self, action, imgs, bus, timeout):
+        self._initial_threads(action, imgs, bus)
+        self._start_threads()
+        self._join_threads(timeout)
+
+    def raise_threads(self):
+        for thread in self._threads:
+            if thread.exc_info:
+                reraise(thread.exc_info[0], thread.exc_info[1], thread.exc_info[2])
+
+    def clean_threads(self):
+        del self.exit_event
+        del self._threads[:]
+
+
+class BlockDevicesPlug(object):
+
+    """
+    The Block Devices Plug.
+    """
+
+    def __init__(self, vm):
+        self.vm = vm
+        self._imgs = vm.params.get("images").split()[1:]
+        self._hotplugged_devs = []
+        self._unplugged_devs = []
+        self._ostype = vm.params['os_type']
+        self._plugged_disks = []
+        self._orig_disks = set()
+        self._all_disks = set()
+        self._lock = threading.Lock()
+        self._qmp_outputs = {}
+        self._event_devs = []
+        self._dev_type = DISK
+        self._qdev_type = qdevices.QBlockdevNode if vm.check_capability(
+            Flags.BLOCKDEV) else qdevices.QDrive
+        self._timeout = 300
+
+    def __getitem__(self, index):
+        """ Get the hot plugged disk index. """
+        if isinstance(index, slice):
+            return [self._plugged_disks[i] for i in range(*index.indices(len(self)))]
+        elif isinstance(index, int):
+            return self._plugged_disks[index]
+        else:
+            raise TestError('PlugManager indices must be integers.')
+
+    def __len__(self):
+        """ Get the len of the hot plugged disks. """
+        return len(self._plugged_disks)
+
+    def __iter__(self):
+        """ Iterate the the hot plugged disks. """
+        for disk in self._plugged_disks:
+            yield disk
+
+    def _list_all_disks(self):
+        """ List all the disks. """
+        session = self.vm.wait_for_login(timeout=360)
+        if self._ostype == "linux":
+            self._all_disks = utils_misc.list_linux_guest_disks(session)
+        else:
+            self._all_disks = set(session.cmd('wmic diskdrive get index').split()[1:])
+        session.close()
+        return self._all_disks
+
+    def _check_qmp_outputs(self, action):
+        """ Check the output of qmp commands. """
+        qmp_outputs = self._qmp_outputs.copy()
+        for dev_id in qmp_outputs:
```

Comment:
this could be like 
```python
for dev_id in list(qmp_outputs.keys()):
```
copy the keys is enough.

Reply:
Thanks, have updated.